File: Sources/LRKeychainWrapper.m
//
//  LRKeychainWrapper.h
//  LoginRadiusSDK
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//
//

#import "LRKeychainWrapper.h"
#import <UIKit/UIKit.h>
#define newCFDict CFDictionaryCreateMutable(kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks)

@interface LRKeychainWrapper ()

@end

@implementation LRKeychainWrapper
static NSData    *publicKeyBits;
static SecKeyRef publicKeyRef;
static SecKeyRef privateKeyRef;

- (instancetype)init {
    NSString *service = [[NSBundle mainBundle] bundleIdentifier];
    return [self initWithIdentifier:service accessGroup:nil];
}

- (instancetype)initWithService:(NSString *)service {
    return [self initWithIdentifier:service accessGroup:nil];
}

- (instancetype)initWithIdentifier:(NSString *)service accessGroup:(NSString *)accessGroup {
    self = [super init];
    if (self) {
        _service = service;
        _accessGroup = accessGroup;
    }
    if ([[[UIDevice currentDevice] systemVersion] floatValue]
        >= 10.0) {
        if(![self lookupPublicKeyRef] || ![self lookupPrivateKeyRef])
            [self generatePasscodeKeyPair];
    }
    
    return self;
}

- (NSData *)dataForKey:(NSString *)key {
    return [self dataForKey:key promptMessage:nil];
}

- (NSData *)dataForKey:(NSString *)key promptMessage:(NSString *)message {
    NSData *d = [self dataForKey:key promptMessage:message error:nil];
    return d;
}

- (NSData *)dataForKey:(NSString *)key promptMessage:(NSString *)message error:(NSError**)err {
    if (!key) {
        return nil;
    }
    
    NSDictionary *query = [self queryFetchOneByKey:key message:message];
    CFTypeRef data = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &data);
    if (status != errSecSuccess) {
        
        return nil;
    }
    
    NSData *dataFound = [NSData dataWithData:(__bridge NSData *)data];
    if (data) {
        CFRelease(data);
    }
    NSData * returnData = dataFound;
    if ([[[UIDevice currentDevice] systemVersion] floatValue]
        >= 10.0)
        returnData = [self decryptData:dataFound];
    return returnData;
}

- (BOOL)hasValueForKey:(NSString *)key {
    if (!key) {
        return NO;
    }
    NSDictionary *query = [self queryFindByKey:key message:nil];
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, NULL);
    return status == errSecSuccess;
}
- (BOOL)setData:(NSData *)data forKey:(NSString *)key {
    return [self setData:data forKey:key promptMessage:nil];
}

- (BOOL)setData:(NSData *)data forKey:(NSString *)key promptMessage:(NSString *)message {
    if (!key) {
        return NO;
    }
    NSData * returnData = data;
    if ([[[UIDevice currentDevice] systemVersion] floatValue]
        >= 10.0)
        returnData = [self encryptData:data];
    NSDictionary *query = [self queryFindByKey:key message:message];
    
    // Normal case
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, NULL);
    if (status == errSecSuccess) {
        if (returnData) {
            NSDictionary *updateQuery = [self queryUpdateValue:returnData message:message];
            
            status = SecItemUpdate((__bridge CFDictionaryRef)query, (__bridge CFDictionaryRef)updateQuery);
            return status == errSecSuccess;
        } else {
            OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
            return status == errSecSuccess;
        }
    } else {
        NSDictionary *newQuery = [self queryNewKey:key value:returnData];
        OSStatus status = SecItemAdd((__bridge CFDictionaryRef)newQuery, NULL);
        return status == errSecSuccess;
    }
}

- (BOOL)deleteEntryForKey:(NSString *)key {
    if (!key) {
        return NO;
    }
    NSDictionary *deleteQuery = [self queryFindByKey:key message:nil];
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)deleteQuery);
    return status == errSecSuccess;
}

- (void)clearAll {
#if !TARGET_IPHONE_SIMULATOR
    NSDictionary *query = [self queryFindAll];
    CFArrayRef result = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef *)&result);
    if (status == errSecSuccess || status == errSecItemNotFound) {
        NSArray *items = [NSArray arrayWithArray:(__bridge NSArray *)result];
        CFBridgingRelease(result);
        for (NSDictionary *item in items) {
            NSMutableDictionary *queryDelete = [[NSMutableDictionary alloc] initWithDictionary:item];
            queryDelete[(__bridge id)kSecClass] = (__bridge id)kSecClassGenericPassword;
            
            OSStatus status = SecItemDelete((__bridge CFDictionaryRef)queryDelete);
            if (status != errSecSuccess) {
                break;
            }
        }
    }
#else
    NSMutableDictionary *queryDelete = [self baseQuery];
    queryDelete[(__bridge id)kSecClass] = (__bridge id)kSecClassGenericPassword;
    //   queryDelete[(__bridge id)kSecUseOperationPrompt] = @"";
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)queryDelete);
    if (status != errSecSuccess) {
        return;
    }
#endif
}

#pragma mark - Query Dictionary Builder methods

- (NSMutableDictionary *)baseQuery {
    NSMutableDictionary *attributes = [@{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: self.service,
    } mutableCopy];
#if !TARGET_IPHONE_SIMULATOR
    if (self.accessGroup) {
        attributes[(__bridge id)kSecAttrAccessGroup] = self.accessGroup;
    }
#endif
    
    return attributes;
}

- (NSDictionary *)queryFindAll {
    NSMutableDictionary *query = [self baseQuery];
    [query addEntriesFromDictionary:@{
        (__bridge id)kSecReturnAttributes: @YES,
        (__bridge id)kSecMatchLimit: (__bridge id)kSecMatchLimitAll,
    }];
    return query;
}

- (NSDictionary *)queryFindByKey:(NSString *)key message:(NSString *)message {
    NSAssert(key != nil, @"Must have a valid non-nil key");
    NSMutableDictionary *query = [self baseQuery];
    query[(__bridge id)kSecAttrAccount] = key;
#if !TARGET_IPHONE_SIMULATOR
    if (message) {
        query[(__bridge id)kSecUseOperationPrompt] = message;
    }
#endif
    return query;
}

- (NSDictionary *)queryUpdateValue:(NSData *)data message:(NSString *)message {
    if (message) {
        return @{
#if !TARGET_IPHONE_SIMULATOR
            (__bridge id)kSecUseOperationPrompt: message,
#endif
            (__bridge id)kSecValueData: data,
        };
    } else {
        return @{
            (__bridge id)kSecValueData: data,
        };
    }
}

- (NSDictionary *)queryNewKey:(NSString *)key value:(NSData *)value {
    NSMutableDictionary *query = [self baseQuery];
    query[(__bridge id)kSecAttrAccount] = key;
    query[(__bridge id)kSecValueData] = value;
    query[(__bridge id)kSecAttrAccessible] = (__bridge id)kSecAttrAccessibleAlwaysThisDeviceOnly;
    
    return query;
}

- (NSDictionary *)queryFetchOneByKey:(NSString *)key message:(NSString *)message {
    NSMutableDictionary *query = [self baseQuery];
    [query addEntriesFromDictionary:@{
        (__bridge id)kSecReturnData: @YES,
        (__bridge id)kSecMatchLimit: (__bridge id)kSecMatchLimitOne,
        (__bridge id)kSecAttrAccount: key,
    }];
    
    return query;
}
- (NSData *)encryptData:(NSData *)data {
    if (data && data.length) {
        
        CFDataRef cipher = SecKeyCreateEncryptedData(publicKeyRef, kSecKeyAlgorithmECIESEncryptionStandardX963SHA256AESGCM, (CFDataRef)data, nil);
        
        // returnData = (__bridge NSData *)cipher;
        return (__bridge NSData *)cipher;
    } else {
        return nil;
        
    }
}
- (NSData*)decryptData:(NSData *)data {
    
    if(data && data.length){
        
        CFDataRef plainData = SecKeyCreateDecryptedData(privateKeyRef, kSecKeyAlgorithmECIESEncryptionStandardX963SHA256AESGCM, (CFDataRef)data, nil);
        return  (__bridge NSData *)plainData;
    }
    else {
        return nil;
    }
}
#pragma mark - Key related methods
- (bool)publicKeyExists
{
    CFTypeRef publicKeyResult = nil;
    CFMutableDictionaryRef publicKeyExistsQuery = newCFDict;
    CFDictionarySetValue(publicKeyExistsQuery, kSecClass,               kSecClassKey);
    CFDictionarySetValue(publicKeyExistsQuery, kSecAttrKeyType,         kSecAttrKeyTypeEC);
    CFDictionarySetValue(publicKeyExistsQuery, kSecAttrApplicationTag,  kPublicKeyName);
    CFDictionarySetValue(publicKeyExistsQuery, kSecAttrKeyClass,        kSecAttrKeyClassPublic);
    CFDictionarySetValue(publicKeyExistsQuery, kSecReturnData,          kCFBooleanTrue);
    
    OSStatus status = SecItemCopyMatching(publicKeyExistsQuery, (CFTypeRef *)&publicKeyResult);
    
    if (status == errSecItemNotFound) {
        return false;
    }
    else if (status == errSecSuccess) {
        return true;
    }
    else {
        [NSException raise:@"Unexpected OSStatus" format:@"Status: %i", status];
        return nil;
    }
}
- (SecKeyRef) lookupPublicKeyRef
{
    OSStatus sanityCheck = noErr;
    NSData *tag;
    id keyClass;
    
    if (publicKeyRef != NULL) {
        // already exists in memory, return
        return publicKeyRef;
    }
    tag = [kPublicKeyName dataUsingEncoding:NSUTF8StringEncoding];
    keyClass = (__bridge id) kSecAttrKeyClassPublic;
    
    
    NSDictionary *queryDict = @{
       
        (__bridge id) kSecClass : (__bridge id) kSecClassKey,
        (__bridge id) kSecAttrKeyType : (__bridge id) kSecAttrKeyTypeEC,
        (__bridge id) kSecAttrApplicationTag : tag,
        (__bridge id) kSecAttrKeyClass : keyClass,
        (__bridge id) kSecReturnRef : (__bridge id) kCFBooleanTrue
    };
    
    sanityCheck = SecItemCopyMatching((__bridge CFDictionaryRef) queryDict, (CFTypeRef *) &publicKeyRef);
    if (sanityCheck != errSecSuccess) {
        NSLog(@"Error trying to retrieve key from server.  sanityCheck: %d", (int)sanityCheck);
    }
    
    
    return publicKeyRef;
}

- (NSData *) publicKeyBits
{
    if (![self publicKeyExists])
        return nil;
    return (NSData *) CFDictionaryGetValue((CFDictionaryRef)[self lookupPublicKeyRef], kSecValueData);
    
}

- (SecKeyRef) lookupPrivateKeyRef
{
    CFMutableDictionaryRef getPrivateKeyRef = newCFDict;
    CFDictionarySetValue(getPrivateKeyRef, kSecClass, kSecClassKey);
    CFDictionarySetValue(getPrivateKeyRef, kSecAttrKeyClass, kSecAttrKeyClassPrivate);
    CFDictionarySetValue(getPrivateKeyRef, kSecAttrLabel, kPrivateKeyName);
    CFDictionarySetValue(getPrivateKeyRef, kSecReturnRef, kCFBooleanTrue);
    //   CFDictionarySetValue(getPrivateKeyRef, kSecUseOperationPrompt, @"Authenticate to sign data");
    
    OSStatus status = SecItemCopyMatching(getPrivateKeyRef, (CFTypeRef *)&privateKeyRef);
    if (status == errSecItemNotFound)
        return nil;
    
    return (SecKeyRef)privateKeyRef;
}

- (bool) generatePasscodeKeyPair
{
    CFErrorRef error = NULL;
    SecAccessControlRef sacObject = SecAccessControlCreateWithFlags(
                                                                    kCFAllocatorDefault,
                                                                    kSecAttrAccessibleAlwaysThisDeviceOnly,
                                                                    kSecAccessControlPrivateKeyUsage,
                                                                    &error
                                                                    );
    
    if (error != errSecSuccess) {
        NSLog(@"Generate key error: %@\n", error);
    }
    
    return [self generateKeyPairWithAccessControlObject:sacObject];
}

- (bool) generateKeyPairWithAccessControlObject:(SecAccessControlRef)accessControlRef
{
    // create dict of private key info
    CFMutableDictionaryRef accessControlDict = newCFDict;;
#if !TARGET_IPHONE_SIMULATOR
    
    CFDictionaryAddValue(accessControlDict, kSecAttrAccessControl, accessControlRef);
#endif
    CFDictionaryAddValue(accessControlDict, kSecAttrIsPermanent, kCFBooleanTrue);
    CFDictionaryAddValue(accessControlDict, kSecAttrLabel, kPrivateKeyName);
    
    // create dict which actually saves key into keychain
    CFMutableDictionaryRef generatePairRef = newCFDict;
#if !TARGET_IPHONE_SIMULATOR
    
    CFDictionaryAddValue(generatePairRef, kSecAttrTokenID, kSecAttrTokenIDSecureEnclave);
#endif
    CFDictionaryAddValue(generatePairRef, kSecAttrKeyType, kSecAttrKeyTypeEC);
    CFDictionaryAddValue(generatePairRef, kSecAttrKeySizeInBits, (__bridge const void *)([NSNumber numberWithInt:256]));
    CFDictionaryAddValue(generatePairRef, kSecPrivateKeyAttrs, accessControlDict);
    
    OSStatus status = SecKeyGeneratePair(generatePairRef, &publicKeyRef, &privateKeyRef);
    
    if (status != errSecSuccess){
        NSLog(@"Error trying to retrieve key from server.  sanityCheck: %d", (int)status);

        return NO;
    }
    [self savePublicKeyFromRef:publicKeyRef];
    return YES;
}

- (bool) savePublicKeyFromRef:(SecKeyRef)publicKeyRef
{   OSStatus sanityCheck = noErr;
    NSData *tag;
    id keyClass;
    
    
    tag = [kPublicKeyName dataUsingEncoding:NSUTF8StringEncoding];
    keyClass = (__bridge id) kSecAttrKeyClassPublic;
    
    
    NSDictionary *saveDict = @{
       
        (__bridge id) kSecClass : (__bridge id) kSecClassKey,
        (__bridge id) kSecAttrKeyType : (__bridge id) kSecAttrKeyTypeEC,
        (__bridge id) kSecAttrApplicationTag : tag,
        (__bridge id) kSecAttrKeyClass : keyClass,
        (__bridge id) kSecValueData : (__bridge NSData *)SecKeyCopyExternalRepresentation(publicKeyRef,nil),
        (__bridge id) kSecAttrKeySizeInBits : [NSNumber numberWithUnsignedInteger:256],
        (__bridge id) kSecAttrEffectiveKeySize : [NSNumber numberWithUnsignedInteger:256],
        (__bridge id) kSecAttrCanDerive : (__bridge id) kCFBooleanFalse,
        (__bridge id) kSecAttrCanEncrypt : (__bridge id) kCFBooleanTrue,
        (__bridge id) kSecAttrCanDecrypt : (__bridge id) kCFBooleanFalse,
        (__bridge id) kSecAttrCanVerify : (__bridge id) kCFBooleanTrue,
        (__bridge id) kSecAttrCanSign : (__bridge id) kCFBooleanFalse,
        (__bridge id) kSecAttrCanWrap : (__bridge id) kCFBooleanTrue,
        (__bridge id) kSecAttrCanUnwrap : (__bridge id) kCFBooleanFalse
    };
    sanityCheck = SecItemAdd((__bridge CFDictionaryRef) saveDict, (CFTypeRef *)&publicKeyRef);
    if (sanityCheck != errSecSuccess) {
        NSLog(@"Error trying to retrieve key from server.  sanityCheck: %d", (int)sanityCheck);

    }
    
    return publicKeyRef;}

- (void) deletePubKey {
    NSDictionary *deleteKeyQuery = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassKey,
        (__bridge id)kSecAttrApplicationTag: [kPublicKeyName dataUsingEncoding:NSUTF8StringEncoding],
        (__bridge id)kSecAttrType: (__bridge id)kSecAttrKeyTypeEC,
    };
    
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)deleteKeyQuery);
   if( status != errSecSuccess)
       NSLog(@"key couldn't be deleted");
}

- (void) deletePrivateKey {
    NSDictionary *deleteKeyQuery = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassKey,
        (__bridge id)kSecAttrApplicationTag: [kPrivateKeyName dataUsingEncoding:NSUTF8StringEncoding],
        (__bridge id)kSecAttrType: (__bridge id)kSecAttrKeyTypeEC,
    };
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)deleteKeyQuery);
      if( status != errSecSuccess)
          NSLog(@"key couldn't be deleted");
}

@end

File: Sources/LRKeychainWrapper.swift
//
//  LRKeychainWrapper.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation
import UIKit
import Security


let PrivateKeyName = "com.loginradius.private"
let PublicKeyName = "com.loginradius.public"
let CFDict = NSMutableDictionary()

/**
 The LRKeychainWrapper class is using SecureEnclave to generate keys and use them to encrypt/decrypt sensitive data.
 It is an abstraction layer for the iPhone Keychain communication and Secure Enclave.
 It has support for sharing keychain items using Access Group and also for iOS 8 fine grained accesibility over a specific Kyechain Item (Using Access Control).
 The SecurrerEnclave support is only available for iOS 10+, otherwise it will default use Keychain.
 
 */
public class LRKeychainWrapper {
    static var publicKeyBits: NSData?
    static var publicKeyRef: SecKey?
    static var privateKeyRef: SecKey?
    
    let service: String
    let accessGroup: String?
    
    public init(service: String, accessGroup: String? = nil) {
        self.service = service
        self.accessGroup = accessGroup
        
        if #available(iOS 10.0, *) {
            if lookupPublicKeyRef() == nil || lookupPrivateKeyRef() == nil {
                generatePasscodeKeyPair()
            }
        }
    }
    
    public convenience init() {
        let service = Bundle.main.bundleIdentifier ?? ""
        self.init(service: service, accessGroup: nil)
    }
    
    public convenience init(service: String) {
        self.init(service: service, accessGroup: nil)
    }
    
    
    public func dataForKey(key: String) -> NSData? {
        return dataForKey(key: key, promptMessage: nil)
    }
    
    
    
    public func dataForKey(key: String, promptMessage: String?) -> NSData? {
        var d:NSData? = self.dataForKey(key: key, promptMessage: promptMessage, error: nil)!
        return d
    }
    
    
    
    public func dataForKey(key: String, promptMessage: String?, error: NSError?) -> NSData? {
        if key.isEmpty {
            return nil
        }
        
        let query = queryFetchOneByKey(key: key, message: promptMessage)
        var data: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &data)
        if status != errSecSuccess {
            return nil
        }
        
        let dataFound = NSData(data: (data as! NSData) as Data)
        if data != nil {
            //CFRelease(data)
        }
        
        
        var returnData = dataFound
        if #available(iOS 10.0, *) {
            returnData = decryptData(data:dataFound)!
        }
        
        
        return returnData
    }
    
    public func hasValueForKey(key: String) -> Bool {
        if key.isEmpty {
            return false
        }
        
        let query = queryFindByKey(key: key, message: nil)
        let status = SecItemCopyMatching(query as CFDictionary, nil)
        return status == errSecSuccess
    }
    
    public func setData(data: NSData, forKey key: String) -> Bool {
        return setData(data: data, forKey: key, promptMessage: nil)
    }
    
    public func setData(data: NSData, forKey key: String, promptMessage: String?) -> Bool {
        if key.isEmpty {
            return false
        }
        
        var returnData = data
        if #available(iOS 10.0, *) {
            returnData = encryptData(data:data)!
        }
        
        let query = queryFindByKey(key: key, message: promptMessage)
        
        var status = SecItemCopyMatching(query as CFDictionary, nil)
        if status == errSecSuccess {
            if returnData != nil {
                let updateQuery = queryUpdateValue(data: returnData as Data, message: promptMessage)
                
                status = SecItemUpdate(query as CFDictionary, updateQuery as CFDictionary)
                return status == errSecSuccess
            } else {
                status = SecItemDelete(query as CFDictionary)
                return status == errSecSuccess
            }
        } else {
            let newQuery = queryNewKey(key: key, value: returnData)
            status = SecItemAdd(newQuery as CFDictionary, nil)
            return status == errSecSuccess
        }
    }
    
    public func deleteEntryForKey(key: String) -> Bool {
        if key.isEmpty {
            return false
        }
        
        let deleteQuery = queryFindByKey(key: key, message: nil)
        let status = SecItemDelete(deleteQuery as CFDictionary)
        return status == errSecSuccess
    }
    
    public func clearAll() {
#if !targetEnvironment(simulator)
        let query = queryFindAll()
        var result: CFTypeRef?
        
        let status = SecItemCopyMatching(query, &result)
        
        if status == errSecSuccess || status == errSecItemNotFound {
            if let items = result as? [NSDictionary] {
                for item in items {
                    var queryDelete = item.mutableCopy() as! NSMutableDictionary
                    queryDelete[kSecClass] = kSecClassGenericPassword
                    
                    let deleteStatus = SecItemDelete(queryDelete as CFDictionary)
                    
                    if deleteStatus != errSecSuccess {
                        break
                    }
                }
            }
        }
#else
        var queryDelete = baseQuery()
        queryDelete[kSecClass] = kSecClassGenericPassword
        
        let status = SecItemDelete(queryDelete as CFDictionary)
        
        if status != errSecSuccess {
            return
        }
#endif
    }
    
    
    // Query Dictionary Builder methods
    
    public func baseQuery() -> NSMutableDictionary {
        var attributes: [AnyHashable: Any] = [
            kSecClass as String: kSecClassGenericPassword as String,
            kSecAttrService as String: self.service
        ]
#if !TARGET_IPHONE_SIMULATOR
        if let accessGroup = self.accessGroup {
            attributes[kSecAttrAccessGroup as String] = accessGroup
        }
#endif
        
        return NSMutableDictionary(dictionary: attributes)
    }
    
    public func queryFindAll() -> NSDictionary {
        var query = baseQuery()
        query[kSecReturnAttributes as String] = kCFBooleanTrue
        query[kSecMatchLimit as String] = kSecMatchLimitAll as String
        return query
    }
    
    public func queryFindByKey(key: String, message: String?) -> NSDictionary {
        assert(key != nil, "Must have a valid non-nil key")
        var query = baseQuery()
        query[kSecAttrAccount as String] = key
#if !TARGET_IPHONE_SIMULATOR
        if let message = message {
            query[kSecUseOperationPrompt as String] = message
        }
#endif
        return query
    }
    
    
    
    public func queryUpdateValue(data: Data, message: String?) -> NSDictionary {
        if let message = message {
            return [
                kSecUseOperationPrompt as String: message,
                kSecValueData as String: data
            ] as NSDictionary
        } else {
            return [
                kSecValueData as String: data
            ] as NSDictionary
        }
    }
    
    public func queryNewKey(key: String, value: NSData) -> NSDictionary {
        var query = baseQuery()
        query[kSecAttrAccount as String] = key
        query[kSecValueData as String] = value
        query[kSecAttrAccessible as String] = kSecAttrAccessibleAlwaysThisDeviceOnly as String
        return query
    }
    
    public func queryFetchOneByKey(key: String, message: String?) -> NSDictionary {
        var query = baseQuery()
        query[kSecReturnData as String] = kCFBooleanTrue
        query[kSecMatchLimit as String] = kSecMatchLimitOne as String
        query[kSecAttrAccount as String] = key
        return query
    }
    
    public func encryptData(data: NSData?) -> NSData? {
        if data != nil && data!.length > 0 {
            let cipher = SecKeyCreateEncryptedData(LRKeychainWrapper.publicKeyRef!, SecKeyAlgorithm.eciesEncryptionStandardX963SHA256AESGCM, data! as CFData, nil)
            return cipher as NSData?
        } else {
            return nil
        }
    }
    
    public func decryptData(data: NSData?) -> NSData? {
        if data != nil && data!.length > 0 {
            let plainData = SecKeyCreateDecryptedData(LRKeychainWrapper.privateKeyRef!, SecKeyAlgorithm.eciesEncryptionStandardX963SHA256AESGCM, data! as CFData, nil)
            return plainData as NSData?
        } else {
            return nil
        }
    }
    
    // MARK: -   Key related methods
    
    
    // Check if public key exists in keychain
    public func publicKeyExists() -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
            kSecReturnAttributes as String: true
        ]
        
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        
        return status == errSecSuccess
    }
    
    // Lookup private key reference in keychain
    public func lookupPrivateKeyRef() -> SecKey? {
        let getPrivateKeyRef = newCFDict
        getPrivateKeyRef.setValue(kSecClassKey, forKey: kSecClass as String)
        getPrivateKeyRef.setValue(kSecAttrKeyClassPrivate, forKey: kSecAttrKeyClass as String)
        getPrivateKeyRef.setValue(PrivateKeyName, forKey: kSecAttrLabel as String)
        getPrivateKeyRef.setValue(kCFBooleanTrue, forKey: kSecReturnRef as String)
        
        var privateKey: CFTypeRef?
        let status = SecItemCopyMatching(getPrivateKeyRef, &privateKey)
        if status == errSecItemNotFound {
            return nil
        }
        
        print("\(privateKey)Private:::::")
        
        return privateKey as! SecKey?
    }
    
    
    // Lookup public key reference in keychain
    public func lookupPublicKeyRef() -> SecKey? {
        if LRKeychainWrapper.publicKeyRef != nil {
            return LRKeychainWrapper.publicKeyRef
        }
        
        let tag = PublicKeyName.data(using: .utf8)
        let keyClass = kSecAttrKeyClassPublic as AnyObject
        
        let queryDict: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeEC,
            kSecAttrApplicationTag: tag!,
            kSecAttrKeyClass: keyClass,
            kSecReturnRef: kCFBooleanTrue
        ]
        
        var publicKey: CFTypeRef?
        let sanityCheck = SecItemCopyMatching(queryDict as CFDictionary, &publicKey)
        if sanityCheck != errSecSuccess {
            print("Error trying to retrieve key from server. sanityCheck: \(sanityCheck)")
        }
        
        return publicKey as! SecKey?
    }
    
    // Generate a passcode key pair
    public func generatePasscodeKeyPair() -> Bool {
        var error: Unmanaged<CFError>?
        let sacObject = SecAccessControlCreateWithFlags(
            kCFAllocatorDefault,
            kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            .privateKeyUsage,
            &error
        )
        
        if error != nil {
            print("Generate key error: \(error!.takeRetainedValue())")
        }
        
        return generateKeyPair(with: sacObject!)
    }
    // Generate a passcode key pair
    public func generateKeyPair(with accessControlRef: SecAccessControl) -> Bool {
        var accessControlDict = newCFDict
#if !TARGET_IPHONE_SIMULATOR
        accessControlDict.setValue(accessControlRef, forKey: kSecAttrAccessControl as String)
#endif
        accessControlDict.setValue(kCFBooleanTrue, forKey: kSecAttrIsPermanent as String)
        accessControlDict.setValue(PrivateKeyName, forKey: kSecAttrLabel as String)
        
        var generatePairRef = newCFDict
#if !TARGET_IPHONE_SIMULATOR
        generatePairRef.setValue(kSecAttrTokenIDSecureEnclave, forKey: kSecAttrTokenID as String)
#endif
        generatePairRef.setValue(kSecAttrKeyTypeEC, forKey: kSecAttrKeyType as String)
        generatePairRef.setValue(256, forKey: kSecAttrKeySizeInBits as String)
        generatePairRef.setValue(accessControlDict, forKey: kSecPrivateKeyAttrs as String)
        
        var publicKey, privateKey: SecKey?
        let status = SecKeyGeneratePair(generatePairRef, &publicKey, &privateKey)
        
        if status != errSecSuccess {
            print("Error trying to retrieve key from server. sanityCheck: \(status)")
            return false
        }
        
        savePublicKey(from: publicKey!)
        return true
    }
    
    // Save public key from reference
    public func savePublicKey(from publicKeyRef: SecKey) -> Bool {
        let tag = PublicKeyName.data(using: .utf8)
        let keyClass = kSecAttrKeyClassPublic as AnyObject
        
        let saveDict: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeEC,
            kSecAttrApplicationTag: tag!,
            kSecAttrKeyClass: keyClass,
            kSecValueData: SecKeyCopyExternalRepresentation(publicKeyRef, nil)!,
            kSecAttrKeySizeInBits: 256,
            kSecAttrEffectiveKeySize: 256,
            kSecAttrCanDerive: kCFBooleanFalse,
            kSecAttrCanEncrypt: kCFBooleanTrue,
            kSecAttrCanDecrypt: kCFBooleanFalse,
            kSecAttrCanVerify: kCFBooleanTrue,
            kSecAttrCanSign: kCFBooleanFalse,
            kSecAttrCanWrap: kCFBooleanTrue,
            kSecAttrCanUnwrap: kCFBooleanFalse
        ]
        
        var sanityCheck: OSStatus = noErr
        sanityCheck = SecItemAdd(saveDict as CFDictionary, nil)
        if sanityCheck != errSecSuccess {
            print("Error trying to retrieve key from server. sanityCheck: \(sanityCheck)")
        }
        
        return publicKeyRef as! Bool
    }
    
    
    
    // Delete public key from keychain
    public func deletePubKey() {
        let deleteKeyQuery: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrApplicationTag: kPublicKeyName.data(using: .utf8)!,
            kSecAttrType: kSecAttrKeyTypeEC
        ]
        
        let status = SecItemDelete(deleteKeyQuery as CFDictionary)
        if status != errSecSuccess {
            print("key couldn't be deleted")
        }
    }
    
    
    
    // Delete private key from keychain
    public func deletePrivateKey() {
        let deleteKeyQuery: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrApplicationTag: kPrivateKeyName.data(using: .utf8)!,
            kSecAttrType: kSecAttrKeyTypeEC
        ]
        
        let status = SecItemDelete(deleteKeyQuery as CFDictionary)
        if status != errSecSuccess {
            print("key couldn't be deleted")
        }
        
    }
}
